daemons/controld/controld_metadata.c

```diff
@@ -354,18 +361,25 @@ controld_get_rsc_metadata(lrm_state_t *lrm_state, const lrmd_rsc_info_t *rsc,
         return NULL;
     }
 
-    /* For now, we always collect resource agent meta-data via a local,
-     * synchronous, direct execution of the agent. This has multiple issues:
-     * the executor should execute agents, not the controller; meta-data for
-     * Pacemaker Remote nodes should be collected on those nodes, not
-     * locally; and the meta-data call shouldn't eat into the timeout of the
-     * real action being performed.
+    /* For most actions, metadata was cached asynchronously before action
+     * execution (via metadata_complete()).
      *
-     * These issues are planned to be addressed by having the scheduler
-     * schedule a meta-data cache check at the beginning of each transition.
-     * Once that is working, this block will only be a fallback in case the
-     * initial collection fails.
+     * However if that failed, and for other actions, retrieve the metadata now
+     * via a local, synchronous, direct execution of the agent.
+     *
+     * This has multiple issues, which is why this is just a fallback: the
+     * executor should execute agents, not the controller; metadata for
+     * Pacemaker Remote nodes should be collected on those nodes, not locally;
+     * the metadata call shouldn't eat into the timeout of the real action being
+     * performed; and the synchronous call blocks the controller (which also
+     * means that if the metadata action tries to contact the controller,
+     * everything will hang until the timeout).
```

Comment:
Another issue is exposed according to what an user experienced. Now given the possibility of having both asynchronous and synchronous actions running at the same time, handling of SIGCHILD is potentially problematic.

According to the symptoms, while an asynchronous action is invoked to retrieve metadata for rsc1 but before it returns,  a synchronous action is invoked to retrieve metadata for rsc2. 

The metadata action for rsc1 times out and results in a zombie process. Based on debugging, the subprocess actually has already exited with status 0, but controller never tries to collect it with waitpid().

According to the code, a synchronous action temporarily blocks SIGCHLD with sigchld_setup() before forking:

https://github.com/ClusterLabs/pacemaker/blob/main/lib/services/services_linux.c#L1219

, and only unblocks it after wait_for_sync_result() returns:

https://github.com/ClusterLabs/pacemaker/blob/main/lib/services/services_linux.c#L1332

AFAIU, it means during the time, SIGCHILD signals from any asynchronous actions are unlikely to be handled by the registered mainloop_signal_handler().

We asked user to try a patch like the following which disables attempts on the synchronous way of retrieving metadata. The issue no long shows up, which seems to confirm the theory:

```
--- a/daemons/controld/controld_cib.c
+++ b/daemons/controld/controld_cib.c
@@ -706,8 +706,7 @@ controld_add_resource_history_xml_as(const char *func, xmlNode *parent,
      * first.
      */
     metadata = controld_get_rsc_metadata(lrm_state, rsc,
-                                         controld_metadata_from_agent
-                                         |controld_metadata_from_cache);
+                                         controld_metadata_from_cache);
     if (metadata == NULL) {
         return;
     }
```
Of course this way, we lose the synchronous fallback. But should we have something imperfect like this as an intermediate solution before switching this over to executor?

Reply:
>> According to the code, a synchronous action temporarily blocks SIGCHLD with sigchld_setup() before forking:
> 
> https://github.com/ClusterLabs/pacemaker/blob/main/lib/services/services_linux.c#L1219
> 
> , and only unblocks it after wait_for_sync_result() returns:
> 
> https://github.com/ClusterLabs/pacemaker/blob/main/lib/services/services_linux.c#L1332
> 
> AFAIU, it means during the time, SIGCHILD signals from any asynchronous actions are unlikely to be handled by the registered mainloop_signal_handler().

It's not really blocking SIGCHLD but redirecting it to a file descriptor for polling instead of the usual handler.

Part of the problem is that `sigchld_received()` assumes any SIGCHLD is for the sync process. struct signalfd_siginfo does have ssi_pid, so we could (and should) ignore SIGCHLD for other PIDs. That way, at least we're waiting until the right child exits. Unfortunately we'd still lose the regular handling for the ignored signals. I don't know if there's any magic with sigqueue() or something to resubmit them after the sync process completes, but even then there would still be the possibility of timing out while the sync process is going. Mixing sync and async children is just a bad idea. :(

I remember the fallback being important, but I forget why. I didn't want to keep it when implementing async metadata, but something needed it.

It might be feasible to have all the callers of `controld_add_resource_history_xml()` ensure that metadata is available before then. However it would be challenging; there are a bunch of paths, and we'd have to find good places backward in the call chains to grab metadata async, and refactor all those to do their stuff in a callback when it's done.

Not sure what the best way forward is ...

Reply:
This was a situation where synchronous way was tried but still failed:

Dec 06 12:05:50.449 node0 pacemaker-controld  [5557] (do_lrm_invoke)        info: Retrieving metadata for db1 (ocf:heartbeat:pgsql) asynchronously
...
Dec 06 12:05:50.465 node0 pacemaker-controld  [5557] (async_action_complete)        debug: db1_meta-data_0[30052] exited with status 1
Dec 06 12:05:50.465 node0 pacemaker-controld  [5557] (do_lrm_rsc_op)        notice: Requesting local execution of monitor operation for db1 on node0 | transition_key=144:69:8:22bd06d2-4da8-4b60-9d4f-a933e0ae458a op_key=db1_monitor_10000
Dec 06 12:05:50.465 node0 pacemaker-controld  [5557] (record_pending_op)    debug: Recording pending op db1_monitor_10000 on node0 in the CIB
Dec 06 12:05:50.465 node0 pacemaker-controld  [5557] (controld_get_rsc_metadata)    debug: Retrieving metadata for db1 (ocf:heartbeat:pgsql) synchronously
Dec 06 12:05:50.481 node0 pacemaker-controld  [5557] (wait_for_sync_result)         info: pgsql_meta-data_0[30057] exited with status 1
Dec 06 12:05:50.481 node0 pacemaker-controld  [5557] (lrmd_api_get_metadata_params)         error: Failed to retrieve meta-data for ocf:heartbeat:pgsql
Dec 06 12:05:50.481 node0 pacemaker-controld  [5557] (controld_get_rsc_metadata)    warning: Failed to get metadata for db1 (ocf:heartbeat:pgsql): Input/output error

Reply:
There may be a reasonable step towards the ultimate fix, that doesn't go as far as executing metadata on remote nodes.

The scheduler could schedule a metadata action (as a normal resource action) before any resource action that requires metadata (`crm_op_needs_metadata()`). For remote actions, the metadata action would be scheduled on the connection host, not the remote node. The processing would be normal as far as graphing and the DC farming them out. When a controller gets a metadata action, it sends it to the executor only if it doesn't already have it cached (with the exception that start actions always require fresh metadata).

Reply:
I separated the smaller project into its own task: https://projects.clusterlabs.org/T770